ocaml/xapi/repository_helpers.ml

```diff
@@ -717,6 +717,224 @@ let get_updates_from_repoquery repositories =
+            ( section
+            , String.starts_with ~prefix:"     replacing " l
+            , String.starts_with ~prefix:"Error: " l
+            )
+          with
+          | Some s, true, false ->
+              (* in a section, but starting with 'replacing', ignoring the line.
+               * https://github.com/rpm-software-management/yum/blob/master/output.py#L1622
+               *)
+              line ~section:(Some s) ~section_acc ~acc
+          | Some s, false, false ->
+              (* in a section, append to the section's list *)
+              line ~section:(Some s) ~section_acc:(l :: section_acc) ~acc
+          | None, false, true ->
+              (* error reported from yum upgrade *)
+              fail l
+          | None, false, false ->
+              (* not in any section, ignoring the line *)
+              line ~section:None ~section_acc:[] ~acc
+          | _ ->
+              fail
+                (Printf.sprintf
+                   "Unexpected output from yum upgrade (dry run): %s" l
+                )
+        )
+      )
+    | true ->
+        return ([], None)
+
+  let word =
+    skip_while is_white >>= fun () ->
+    take_while1 (fun x -> not (is_white x)) >>= fun s ->
+    skip_while is_white >>= fun () -> return s
+
+  let package =
+    word >>= fun name ->
+    word >>= fun arch ->
+    word >>= fun epoch_version_release ->
+    word >>= fun repo ->
+    word >>= fun _ ->
+    word >>= fun _ ->
+    match Pkg.parse_epoch_version_release epoch_version_release with
+    | epoch, version, release ->
+        return (Pkg.{name; epoch; version; release; arch}, repo)
+    | exception e ->
+        let msg =
+          Printf.sprintf "parse_epoch_version_release: %s"
+            (ExnHelper.string_of_exn e)
+        in
+        fail msg
+
+  let packages = many_till package end_of_input
+
+  let parse_yum_txn_file txn_line_opt =
+    let txn_line =
+      skip_while is_white >>= fun () ->
+      string "yum load-transaction " >>= fun _ ->
+      take_till is_white >>= fun s -> return s
+    in
+    match txn_line_opt with
+    | Some line -> (
+      match parse_string ~consume txn_line line with
+      | Ok txn_file_path ->
+          Some txn_file_path
+      | Error msg ->
+          warn "Can't parse transaction file from line %s: %s." line msg ;
+          None
+    )
+    | None ->
+        None
+
+  let parse_output_of_dry_run output =
+    Astring.String.cuts ~sep:"\n" output
+    |> List.iter (fun x -> debug "yum upgrade (dry run): %s" x) ;
+
+    match
+      (parse_string ~consume (line ~section:None ~section_acc:[] ~acc:[]))
+        output
+    with
+    | Ok ((_ :: _ as sections), txn_line_opt) -> (
+        let txn_file = parse_yum_txn_file txn_line_opt in
+        let get_pkgs lines =
+          let s = String.concat "\n" (List.rev_append lines []) in
+          parse_string ~consume packages s
+        in
+        sections
+        |> List.filter (fun (section, _) ->
+               [
```

Comment:
I would write the filter function as 
```
fun (section, _) -> match section with 
|  "Installing" -> true
| "Updating" -> true
| ..
| _ -> false
```
and avoid building a list and using `List.mem`